### PR TITLE
CSP-928 AAN Apprentice Directory Message Employer Web

### PR DESCRIPTION
**Pull request by ManishAgarwalDfE**

### CODE REVIEW

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs [hunk 1]
```diff
@@ -171,4 +177,156 @@ public void MemberProfileDetailMapping_ReturnsMemberProfileDetail(MemberUserType
...
         });
     }
+
+    [Test]
+    [MoqInlineAutoData(MemberUserType.Apprentice)]
+    [MoqInlineAutoData(MemberUserType.Employer)]
+    public async Task Post_InvalidCommand_ReturnsMemberProfileView(
+       MemberUserType userType,
+       SubmitConnectionCommand command,
+       [Frozen] Mock<IOuterApiClient> outerApiMock,
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs:187]: [Frozen] is pointless here

File: src/SFA.DAS.Employer.Aan.Web/wwwroot/assets/PDF-AAN-handbook-2022.pdf
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web/wwwroot/assets/PDF-AAN-handbook-2022.pdf:1]: Why is this file added here?

File: src/SFA.DAS.Employer.Aan.Web/Controllers/MemberProfileController.cs [hunk 2]
```diff
@@ -83,45 +88,83 @@ public async Task<IActionResult> Get([FromRoute] string employerAccountId, [From
...
+            return View(MemberProfileViewPath, model);
+        }
+        CreateNotificationRequest createNotificationRequest = new CreateNotificationRequest(id, command.ReasonToGetInTouch);
+        var response = await _outerApiClient.PostNotification(memberId, createNotificationRequest, cancellationToken);
+
+        if (response.ResponseMessage.IsSuccessStatusCode)
+        {
+            return RedirectToAction(SharedRouteNames.NotificationSentConfirmation, new { employerAccountId });
+        }
+        throw new InvalidOperationException($"A problem occured while sending notification.");
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web/Controllers/MemberProfileController.cs:118]: This can be simplified by not allowing any status codes in the outer api. Which means that if success code is not returned an exception will be raised at the infra level.

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+            CodeOfConduct = true,
+            DetailShareAllowed = true
+        };
+
+        //Act
+        var sut = new SubmitConnectionCommandValidator();
+        var result = sut.TestValidate(model);
+
+        //Assert
+        result.ShouldNotHaveValidationErrorFor("CodeOfConduct");
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs:124]: Use expression instead of string

Comment (DevenKShah) [reply src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs:124]: I thinks you need to verify no errors on DetailsShareAllowed proeprty

File: src/SFA.DAS.Employer.Aan.Domain/Interfaces/IOuterApiClient.cs [hunk 4]
```diff
@@ -58,4 +58,7 @@ public interface IOuterApiClient
     [Get("/members/{memberId}/profile")]
     [AllowAnyStatusCode]
     Task<Response<GetMemberProfileResponse>> GetMemberProfile([Header(RequestHeaders.RequestedByMemberIdHeader)] Guid requestedByMemberId, [Path] Guid memberId, [Query] bool @public, CancellationToken cancellationToken);
+
+    [Post("/notifications")]
+    Task<Response<CreateNotificationResponse>> PostNotification([Header(RequestHeaders.RequestedByMemberIdHeader)] Guid requestedByMemberId, [Body] CreateNotificationRequest request, CancellationToken cancellationToken);
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Domain/Interfaces/IOuterApiClient.cs:63]: Remove generic Response from result type.

File: src/SFA.DAS.Employer.Aan.Web/Validators/MemberProfile/SubmitConnectionCommandValidator.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    public const string ReasonToConnectValidationMessage = "You must tell us why you want to get in touch";
+    public const string AcceptDetailShareAllowedMessage = "You must agree to sharing your details";
+    public const string AcceptCodeOfConductMessage = "You must agree to adhere the code of conduct";
+
+    public SubmitConnectionCommandValidator()
+    {
+        RuleLevelCascadeMode = CascadeMode.Stop;
+        RuleFor(x => x.ReasonToGetInTouch).GreaterThan(0).WithMessage(ReasonToConnectValidationMessage);
+        RuleFor(x => x.DetailShareAllowed).Must(beTrue => beTrue).WithMessage(AcceptDetailShareAllowedMessage);
+        RuleFor(x => x.CodeOfConduct).Must(beTrue => beTrue).WithMessage(AcceptCodeOfConductMessage);
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web/Validators/MemberProfile/SubmitConnectionCommandValidator.cs:17]: Booleans should always read as a question. These properties needs renaming. HasAgreedToSharePersonalDetails and HasAgreedToCodeOfConduct.

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs:85]: Use expression

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            CodeOfConduct = true,
+            DetailShareAllowed = true
+        };
+
+        //Act
+        var sut = new SubmitConnectionCommandValidator();
+        var result = sut.TestValidate(model);
+
+        //Assert
+        result.ShouldNotHaveValidationErrorFor("ReasonToGetInTouch");
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs:46]: Use expression

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs [hunk 8]
```diff
@@ -171,4 +177,156 @@ public void MemberProfileDetailMapping_ReturnsMemberProfileDetail(MemberUserType
...
+        sessionServiceMock.Setup(s => s.Get(Constants.SessionKeys.MemberId)).Returns(memberId.ToString());
+        var user = UsersForTesting.GetUserWithClaims(employerId);
+        var response = new Response<GetMemberProfileResponse>(string.Empty, new(HttpStatusCode.OK), () => getMemberProfileResponse);
+        outerApiMock.Setup(o => o.GetMemberProfile(memberId, memberId, It.IsAny<bool>(), It.IsAny<CancellationToken>()))
+                    .Returns(Task.FromResult(response));
+        MemberProfileController sut = new MemberProfileController(outerApiMock.Object, sessionServiceMock.Object, validatorMock.Object);
+        sut.ControllerContext = new ControllerContext { HttpContext = new DefaultHttpContext { User = user } };
+
+        //Act
+        var result = await sut.Post(employerId, memberId, command, cancellationToken);
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs:207]: I do not see where the validation errors are raised here?

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs [hunk 9]
```diff
@@ -171,4 +177,156 @@ public void MemberProfileDetailMapping_ReturnsMemberProfileDetail(MemberUserType
...
+
+        //Assert
+        Assert.That(result, Is.InstanceOf<ViewResult>());
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs:210]: This assert is not enough and is misleading. You should check for ViewName at least.

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs:1]: Could have used a separate file for Post tests

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs [hunk 10]
```diff
@@ -171,4 +177,156 @@ public void MemberProfileDetailMapping_ReturnsMemberProfileDetail(MemberUserType
...
+
+        //Act
+        var result = await sut.Post(employerId, memberId, command, cancellationToken);
+
+        //Assert
+        Assert.Multiple(() =>
+        {
+            Assert.That(result, Is.TypeOf<RedirectToActionResult>());
+            var redirectToAction = (RedirectToActionResult)result;
+            Assert.That(redirectToAction.ActionName, Is.EqualTo("NotificationSentConfirmation"));
```
Comment (DevenKShah) [inline src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs:256]: Use nameof so refactoring becomes easy.

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs (hunk 1 above)
Comment (ManishAgarwalDfE) [reply src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs:187]: Done

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs (hunk 3 above)
Comment (ManishAgarwalDfE) [reply src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs:124]: Done

File: src/SFA.DAS.Employer.Aan.Web/Controllers/MemberProfileController.cs (hunk 2 above)
Comment (ManishAgarwalDfE) [reply src/SFA.DAS.Employer.Aan.Web/Controllers/MemberProfileController.cs:118]: Done

File: src/SFA.DAS.Employer.Aan.Web/Validators/MemberProfile/SubmitConnectionCommandValidator.cs (hunk 5 above)
Comment (ManishAgarwalDfE) [reply src/SFA.DAS.Employer.Aan.Web/Validators/MemberProfile/SubmitConnectionCommandValidator.cs:17]: I have created a task for this and cover in a separate ticket;https://skillsfundingagency.atlassian.net.mcas.ms/browse/CSP-1118

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs (hunk 7 above)
Comment (ManishAgarwalDfE) [reply src/SFA.DAS.Employer.Aan.Web.UnitTests/Validators/MemberProfile/SubmitConnectionCommandValidatorTests.cs:46]: Done

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs (hunk 9 above)
Comment (ManishAgarwalDfE) [reply src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs:210]: Done

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs
Comment (ManishAgarwalDfE) [reply src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs:1]: Done

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs (hunk 10 above)
Comment (ManishAgarwalDfE) [reply src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs:256]: Done

File: src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs (hunk 8 above)
Comment (ManishAgarwalDfE) [reply src/SFA.DAS.Employer.Aan.Web.UnitTests/Controllers/MemberProfileControllerTests.cs]: The validation fails here; command.ReasonToGetInTouch = 0; This renders ReasonToGetInTouch to have a value less than 1

**Review by MarkFCain — DISMISSED** [review]

**Review by MarkFCain — APPROVED** [review]
